diff --git a/css/doc-explorer.css b/css/doc-explorer.css
index 7406255..c169db4 100644
--- a/css/doc-explorer.css
+++ b/css/doc-explorer.css
@@ -72,6 +72,20 @@
   min-width: 300px;
 }
 
+.graphiql-container .doc-type-kind {
+  border-bottom: 1px solid #e0e0e0;
+  color: #777;
+  cursor: default;
+  font-size: 20px;
+  font-variant: small-caps;
+  font-weight: bold;
+  letter-spacing: 1px;
+  margin: 0 -15px 10px 0;
+  padding: 10px 0;
+  -webkit-user-select: none;
+          user-select: none;
+}
+
 .graphiql-container .doc-type-description p:first-child ,
 .graphiql-container .doc-type-description blockquote:first-child {
   margin-top: 0;
diff --git a/package.json b/package.json
index 5712c69..9e9c69e 100644
--- a/package.json
+++ b/package.json
@@ -49,8 +49,8 @@
     ]
   },
   "dependencies": {
-    "codemirror": "^5.26.0",
-    "codemirror-graphql": "^0.6.8",
+    "codemirror": "^5.27.4",
+    "codemirror-graphql": "^0.6.9",
     "marked": "0.3.6"
   },
   "peerDependencies": {
@@ -60,12 +60,12 @@
     "react-dom": "^15.6.0"
   },
   "devDependencies": {
-    "autoprefixer": "^7.0.0",
+    "autoprefixer": "^7.1.1",
     "babel-cli": "6.24.1",
     "babel-eslint": "7.2.3",
     "babel-plugin-syntax-async-functions": "6.13.0",
     "babel-plugin-transform-class-properties": "6.24.1",
-    "babel-plugin-transform-object-rest-spread": "6.22.0",
+    "babel-plugin-transform-object-rest-spread": "6.23.0",
     "babel-plugin-transform-regenerator": "6.24.1",
     "babel-preset-es2015": "6.24.1",
     "babel-preset-react": "6.24.1",
@@ -80,19 +80,20 @@
     "express": "5.0.0-alpha.5",
     "express-graphql": "0.6.6",
     "flow-bin": "0.49.1",
-    "graphql": "^0.10.1",
-    "husky": "^0.14.0",
+    "graphql": "^0.10.3",
+    "husky": "^0.14.2",
     "jsdom": "11.0.0",
     "lint-staged": "^4.0.0",
     "mocha": "3.4.2",
     "postcss-cli": "4.1.0",
-    "prettier": "^1.4.4",
-    "prop-types": "15.5.8",
-    "react": "15.5.4",
-    "react-dom": "15.5.4",
+    "prettier": "^1.5.2",
+    "prop-types": "15.5.10",
+    "react": "15.6.1",
+    "react-dom": "15.6.1",
     "react-test-renderer": "15.6.1",
-    "uglify-js": "^3.0.20",
+    "uglify-js": "^3.0.22",
     "uglifyify": "^4.0.2",
+    "html-to-react": "1.2.11",
     "watchify": "3.9.0"
   }
 }
diff --git a/src/components/DocExplorer.js b/src/components/DocExplorer.js
index 6487db2..3d1c328 100644
--- a/src/components/DocExplorer.js
+++ b/src/components/DocExplorer.js
@@ -102,6 +102,7 @@ export class DocExplorer extends React.Component {
         <FieldDoc
           field={navItem.def}
           onClickType={this.handleClickTypeOrField}
+          schema={schema}
         />;
     }
 
diff --git a/src/components/DocExplorer/FieldDoc.js b/src/components/DocExplorer/FieldDoc.js
index f425168..43b287c 100644
--- a/src/components/DocExplorer/FieldDoc.js
+++ b/src/components/DocExplorer/FieldDoc.js
@@ -12,9 +12,11 @@ import PropTypes from 'prop-types';
 import Argument from './Argument';
 import MarkdownContent from './MarkdownContent';
 import TypeLink from './TypeLink';
+import { GraphQLSchema } from 'graphql';
 
 export default class FieldDoc extends React.Component {
   static propTypes = {
+    schema: PropTypes.instanceOf(GraphQLSchema),
     field: PropTypes.object,
     onClickType: PropTypes.func,
   }
@@ -39,6 +41,8 @@ export default class FieldDoc extends React.Component {
                 <Argument arg={arg} onClickType={this.props.onClickType} />
               </div>
               <MarkdownContent
+                schema={this.props.schema}
+                onClickType={this.props.onClickType}
                 className="doc-value-description"
                 markdown={arg.description}
               />
@@ -50,13 +54,20 @@ export default class FieldDoc extends React.Component {
 
     return (
       <div>
+        <div className="doc-type-kind">
+          {'field'}
+        </div>
         <MarkdownContent
+          schema={this.props.schema}
+          onClickType={this.props.onClickType}
           className="doc-type-description"
           markdown={field.description || 'No Description'}
         />
         {
           field.deprecationReason &&
           <MarkdownContent
+            schema={this.props.schema}
+            onClickType={this.props.onClickType}
             className="doc-deprecation"
             markdown={field.deprecationReason}
           />
diff --git a/src/components/DocExplorer/MarkdownContent.js b/src/components/DocExplorer/MarkdownContent.js
index 12f4cad..a8602d6 100644
--- a/src/components/DocExplorer/MarkdownContent.js
+++ b/src/components/DocExplorer/MarkdownContent.js
@@ -9,11 +9,16 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 import Marked from 'marked';
+import { GraphQLSchema } from 'graphql';
+import TypeLink from './TypeLink';
+import htmlToReact from 'html-to-react';
 
 export default class MarkdownContent extends React.Component {
   static propTypes = {
     markdown: PropTypes.string,
     className: PropTypes.string,
+    schema: PropTypes.instanceOf(GraphQLSchema),
+    onClickType: PropTypes.func
   }
 
   shouldComponentUpdate(nextProps) {
@@ -26,12 +31,33 @@ export default class MarkdownContent extends React.Component {
       return <div />;
     }
 
-    const html = Marked(markdown, { sanitize: true });
-    return (
-      <div
-        className={this.props.className}
-        dangerouslySetInnerHTML={{ __html: html }}
-      />
-    );
+    let html = Marked(markdown, { sanitize: true });
+
+    html = `<div class="${this.props.className}">${html}</div>`
+
+    let typeMap = this.props.schema.getTypeMap();
+    let processNodeDefinitions = new htmlToReact.ProcessNodeDefinitions(React);
+    let instructions = [
+        {
+            shouldProcessNode: (node) => {
+                return node && node.name && node.name === "a";
+            },
+            processNode: (node, children, index) => {
+                let name = node.children.map((child) => child.data).join("");
+                let type = typeMap[name]
+                if (type === undefined)
+                    return processNodeDefinitions.processDefaultNode(node, children, index)
+                else
+                    return <TypeLink type={type} onClick={this.props.onClickType} />;
+            }
+        }, {
+            shouldProcessNode: (node) => true,
+            processNode: processNodeDefinitions.processDefaultNode
+        }
+    ];
+    var htmlToReactParser = new htmlToReact.Parser();
+    var markup = htmlToReactParser.parseWithInstructions(html, () => true, instructions);
+
+    return markup;
   }
 }
diff --git a/src/components/DocExplorer/SchemaDoc.js b/src/components/DocExplorer/SchemaDoc.js
index 6aa6bb3..fd3c5ac 100644
--- a/src/components/DocExplorer/SchemaDoc.js
+++ b/src/components/DocExplorer/SchemaDoc.js
@@ -33,6 +33,8 @@ export default class SchemaDoc extends React.Component {
     return (
       <div>
         <MarkdownContent
+          schema={schema}
+          onClickType={this.props.onClickType}
           className="doc-type-description"
           markdown={
             'A GraphQL schema provides a root type for each kind of operation.'
diff --git a/src/components/DocExplorer/TypeDoc.js b/src/components/DocExplorer/TypeDoc.js
index 80d81e6..fd0fdad 100644
--- a/src/components/DocExplorer/TypeDoc.js
+++ b/src/components/DocExplorer/TypeDoc.js
@@ -157,7 +157,12 @@ export default class TypeDoc extends React.Component {
 
     return (
       <div>
+        <div className="doc-type-kind">
+          {'type'}
+        </div>
         <MarkdownContent
+          schema={this.props.schema}
+          onClickType={this.props.onClickType}
           className="doc-type-description"
           markdown={type.description || 'No Description'}
         />
@@ -204,6 +209,8 @@ function Field({ type, field, onClickType, onClickField }) {
       {
         field.deprecationReason &&
         <MarkdownContent
+          schema={this.props.schema}
+          onClickType={this.props.onClickType}
           className="doc-deprecation"
           markdown={field.deprecationReason}
         />
@@ -226,12 +233,16 @@ function EnumValue({ value }) {
         {value.name}
       </div>
       <MarkdownContent
+        schema={this.props.schema}
+        onClickType={this.props.onClickType}
         className="doc-value-description"
         markdown={value.description}
       />
       {
         value.deprecationReason &&
         <MarkdownContent
+          schema={this.props.schema}
+          onClickType={this.props.onClickType}
           className="doc-deprecation"
           markdown={value.deprecationReason}
         />
diff --git a/src/components/GraphiQL.js b/src/components/GraphiQL.js
index 34a565d..604baf5 100644
--- a/src/components/GraphiQL.js
+++ b/src/components/GraphiQL.js
@@ -664,7 +664,10 @@ export class GraphiQL extends React.Component {
 
   handlePrettifyQuery = () => {
     const editor = this.getQueryEditor();
-    editor.setValue(print(parse(editor.getValue())));
+    editor.setValue(print(parse(editor.getValue()).replace(/^( +)/mg, "$1$1")));
+    const variables = JSON.stringify(JSON.parse(this.state.variables), null, 4);
+    const varEditor = this.variableEditorComponent.getCodeMirror();
+    varEditor.setValue(variables);
   }
 
   handleEditQuery = debounce(100, value => {
